streams/src/test/java/org/apache/kafka/streams/processor/TopologyBuilderTest.java

```diff
@@ -50,9 +52,95 @@
+
+        final String earliestTopic = "earliestTopic";
+        final String latestTopic = "latestTopic";
+
+        builder.addSource(TopologyBuilder.AutoOffsetReset.EARLIEST, "source", earliestTopic);
+        builder.addSource(TopologyBuilder.AutoOffsetReset.LATEST, "source2", latestTopic);
+
+        assertTrue(builder.earliestResetTopicsPattern().matcher(earliestTopic).matches());
+        assertTrue(builder.latestResetTopicsPattern().matcher(latestTopic).matches());
+
+    }
+
+    @Test
+    public void shouldAddSourcePatternWithOffsetReset() {
+        final TopologyBuilder builder = new TopologyBuilder();
+        
+        final String earliestTopicPattern = "earliest.*Topic";
+        final String latestTopicPattern = "latest.*Topic";
+
+        builder.addSource(TopologyBuilder.AutoOffsetReset.EARLIEST, "source", Pattern.compile(earliestTopicPattern));
+        builder.addSource(TopologyBuilder.AutoOffsetReset.LATEST, "source2", Pattern.compile(latestTopicPattern));
+
+        assertTrue(builder.earliestResetTopicsPattern().matcher("earliestTestTopic").matches());
+        assertTrue(builder.latestResetTopicsPattern().matcher("latestTestTopic").matches());
+    }
+
+    @Test
+    public void shouldAddSourceWithoutOffsetReset() {
+        final TopologyBuilder builder = new TopologyBuilder();
+        final Serde<String> stringSerde = Serdes.String();
+
+        final Pattern expectedPattern = Pattern.compile("test-topic");
+
+        builder.addSource("source", stringSerde.deserializer(), stringSerde.deserializer(), "test-topic");
+
+
+
+        assertEquals(expectedPattern.pattern(), builder.sourceTopicPattern().pattern());
+
+        assertEquals(builder.earliestResetTopicsPattern().pattern(), "");
+        assertEquals(builder.latestResetTopicsPattern().pattern(), "");
+    }
+
+    @Test
+    public void shouldAddPatternSourceWithoutOffsetReset() {
+        final TopologyBuilder builder = new TopologyBuilder();
+        final Serde<String> stringSerde = Serdes.String();
+
+        final Pattern expectedPattern = Pattern.compile("test-.*");
+        
+        builder.addSource("source", stringSerde.deserializer(), stringSerde.deserializer(), Pattern.compile("test-.*"));
+
+
+
+        assertEquals(expectedPattern.pattern(), builder.sourceTopicPattern().pattern());
+
+        assertEquals(builder.earliestResetTopicsPattern().pattern(), "");
+        assertEquals(builder.latestResetTopicsPattern().pattern(), "");
+    }
+
+    @Test(expected = TopologyBuilderException.class)
+    public void shouldNotAllowOffsetResetSourceWithoutTopics() {
+        final TopologyBuilder builder = new TopologyBuilder();
+        final Serde<String> stringSerde = Serdes.String();
+
+        builder.addSource(TopologyBuilder.AutoOffsetReset.EARLIEST, "source", stringSerde.deserializer(), stringSerde.deserializer(), new String[]{});
+        fail("Should throw TopologyBuilderException with no topics");
```

Comment:
`fail` is not required.

Maybe, it would be better though to have a try-catch around this (and use `fail`) and remove `expected` annoation (using `expected` should only be done for single-line tests).